bitcoin/src/blockdata/fee_rate.rs

```diff
@@ -4,102 +4,163 @@ use core::fmt;
+pub struct FeeRate {
+    per_kwu: Amount,
+}
 
 impl FeeRate {
     /// 0 sat/kwu.
     ///
     /// Equivalent to [`MIN`](Self::MIN), may better express intent in some contexts.
-    pub const ZERO: FeeRate = FeeRate(0);
+    pub const ZERO: FeeRate = FeeRate { per_kwu: Amount::ZERO };
 
     /// Minimum possible value (0 sat/kwu).
     ///
     /// Equivalent to [`ZERO`](Self::ZERO), may better express intent in some contexts.
-    pub const MIN: FeeRate = FeeRate(u64::min_value());
+    pub const MIN: FeeRate = FeeRate { per_kwu: Amount::min_value() };
 
     /// Maximum possible value.
-    pub const MAX: FeeRate = FeeRate(u64::max_value());
+    pub const MAX: FeeRate = FeeRate { per_kwu: Amount::max_value() };
 
     /// Minimum fee rate required to broadcast a transaction.
     ///
     /// The value matches the default Bitcoin Core policy at the time of library release.
-    pub const BROADCAST_MIN: FeeRate = FeeRate::from_sat_per_vb_unchecked(1);
+    pub const BROADCAST_MIN: FeeRate = FeeRate::from_per_vb_unchecked(Amount::from_sat(1));
 
     /// Fee rate used to compute dust amount.
-    pub const DUST: FeeRate = FeeRate::from_sat_per_vb_unchecked(3);
+    pub const DUST: FeeRate = FeeRate::from_per_vb_unchecked(Amount::from_sat(3));
 
-    /// Constructs `FeeRate` from satoshis per 1000 weight units.
-    pub const fn from_sat_per_kwu(sat_kwu: u64) -> Self { FeeRate(sat_kwu) }
+    /// Constructs `FeeRate` from an amount per 1000 weight units.
+    pub const fn from_per_kwu(per_kwu: Amount) -> Self { FeeRate { per_kwu } }
 
-    /// Constructs `FeeRate` from satoshis per virtual bytes.
+    /// Constructs `FeeRate` from an amount per virtual bytes.
     ///
     /// # Errors
     ///
     /// Returns `None` on arithmetic overflow.
-    pub fn from_sat_per_vb(sat_vb: u64) -> Option<Self> {
+    pub const fn from_per_vb(per_vb: Amount) -> Option<Self> {
         // 1 vb == 4 wu
         // 1 sat/vb == 1/4 sat/wu
         // sat_vb sat/vb * 1000 / 4 == sat/kwu
-        Some(FeeRate(sat_vb.checked_mul(1000 / 4)?))
+
+        // NB remove indirection after impl const support for ops
+        if let Some(per_kwu) = per_vb.checked_mul(1000 / 4) {
+            Some(FeeRate { per_kwu })
+        } else {
+            None
+        }
     }
 
-    /// Constructs `FeeRate` from satoshis per virtual bytes without overflow check.
-    pub const fn from_sat_per_vb_unchecked(sat_vb: u64) -> Self { FeeRate(sat_vb * (1000 / 4)) }
+    /// Constructs `FeeRate` from an amount per virtual bytes without overflow check.
+    pub const fn from_per_vb_unchecked(per_vb: Amount) -> Self {
+        FeeRate { per_kwu: Amount::from_sat(per_vb.to_sat() * (1000 / 4)) }
+    }
 
-    /// Returns raw fee rate.
-    ///
-    /// Can be used instead of `into()` to avoid inference issues.
-    pub const fn to_sat_per_kwu(self) -> u64 { self.0 }
+    /// Returns the fee rate per kilo weight unit.
+    pub const fn to_per_kwu(self) -> Amount { self.per_kwu }
```

Comment:
This function name feels confusing.  If it's returning an `Amount` type now, it's no longer a ratio of fee to size (fee_rate).  I feel like `to_amount()` or something makes more sense.

Reply:
More like `to_amount_per_kwu`. It still is ratio but expressed in confusing form. I think just dividing by 1kwu would be much clearer.

Reply:
Suppose we renamed the function `times_kwu`?

Reply:
It's not cast, it's proper conversion since it explicitly says `per_kwu`. But your confusion proves my point: this is confusing. :)

Reply:
err yeah sorry, type cast isn't the right term.

Reply:
@apoelstra WTF I didn't see your comment. I find that even more confusing. I don't understand why people can't just write `fee_rate * Weight::ONE_KWU`.

Reply:
Oh, yeah, I'm happy with that. So we should just remove this method.

Reply:
Well for one multiplications aren't const yet. And also they are way harder to discover for users. I don't see any problem with having some cheap conversions to and from some standard weight/size values.

Reply:
Ah, `const` is a good argument. But could it be named `to_amount_per_kwu`? Seems clearer to me.

Reply:
Sure, wouldn't object if others also find that clearer. To me it seems verbose as if you "fix" a "feerate" to "per kwu" what remains is an "amount" automatically.